### PR TITLE
DEV: replace diffhtml with morphlex

### PR DESCRIPTION
**Pull request by ZogStriP**
We've been having some performance issues with diffhtml and are trying out this newer & smaller library called morphlex.

No tests were changed since this should be a drop-in replacement (at least behaviour-wise).

### CODE REVIEW

File: lib/tasks/javascript.rake [hunk 1]
```diff
@@ -74,6 +74,7 @@ def dependencies
     { source: "chart.js/dist/chart.min.js", public: true },
     { source: "chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.min.js", public: true },
     { source: "diffhtml/dist/diffhtml.min.js", public: true },
```
Comment (pmusaraj) [inline lib/tasks/javascript.rake:76]: We should probably remove diffhtml here, yes?

File: lib/tasks/javascript.rake [hunk 2]
```diff
@@ -74,6 +74,7 @@ def dependencies
...
+    { source: "morphlex/dist/morphlex.js", public: true },
```
Comment (pmusaraj) [inline lib/tasks/javascript.rake:77]: I think the "new" way of adding dependencies is via webpack. I think we did `jsuites` that way recently, see https://github.com/discourse/discourse/pull/25018

**Review by pmusaraj — COMMENTED** [review]
This looks promising, looking forward to trying it out!

File: lib/tasks/javascript.rake (hunk 2 above)
Comment (ZogStriP) [reply lib/tasks/javascript.rake:77]: Thanks 👍

Comment (davidtaylorhq) [reply lib/tasks/javascript.rake:77]: Yup exactly 💯 

Bonus points for using async `import()` so that it only gets loaded when the composer is opened!

Comment (ZogStriP) [reply lib/tasks/javascript.rake]: It's still a WIP, but is that what you had in mind when you said "async `import()`" @davidtaylorhq ?

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 3]
```diff
@@ -413,92 +405,81 @@ export default Component.extend(TextareaTextManipulation, {
...
+      const { innerHTML } = await import("diffhtml");
+      innerHTML(this._previewElement, cookedElement.innerHTML);
 
-          if (this.previewUpdated) {
-            this.previewUpdated(preview);
-          }
-        });
-      });
+      // const { morph } = await import("morphlex");
+      // morph(this._previewElement, cookedElement);
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:460]: I can easily switch between `diffhtml` and `morphlex` now 🙌

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 4]
```diff
@@ -231,6 +238,8 @@ export default Component.extend(TextareaTextManipulation, {
   ready: false,
   lastSel: null,
   _itsatrap: null,
+  _previewElement: null,
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:241]: Caches the preview `textarea` so we don't query the DOM every single time.

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 5]
```diff
@@ -231,6 +238,8 @@ export default Component.extend(TextareaTextManipulation, {
...
+  _morphFunction: null,
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:242]: Caches the morphing function so it's only loaded once instead of on every (debounced) keystrokes.

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 6]
```diff
@@ -260,19 +269,12 @@ export default Component.extend(TextareaTextManipulation, {
...
+    return formTemplateIds?.length > 0 && !replyingToTopic && !editingPost;
   },
 
   @discourseComputed("placeholder")
   placeholderTranslated(placeholder) {
-    if (placeholder) {
-      return I18n.t(placeholder);
-    }
-    return null;
+    return placeholder ? I18n.t(placeholder) : null;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:277]: Unrelated refactoring.

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 7]
```diff
@@ -319,20 +322,15 @@ export default Component.extend(TextareaTextManipulation, {
...
-      this.appEvents.on("composer:insert-block", this, "insertBlock");
-      this.appEvents.on("composer:insert-text", this, "insertText");
-      this.appEvents.on("composer:replace-text", this, "replaceText");
-      this.appEvents.on("composer:apply-surround", this, "_applySurround");
-      this.appEvents.on(
-        "composer:indent-selected-text",
-        this,
-        "indentSelection"
-      );
+      for (const [event, handler] of COMPOSER_EVENTS) {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:331]: Unrelated refactoring.

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 8]
```diff
@@ -413,92 +407,82 @@ export default Component.extend(TextareaTextManipulation, {
     return toolbar;
   },
 
-  cachedCookAsync(text) {
-    if (this._cachedCookFunction) {
-      return Promise.resolve(this._cachedCookFunction(text));
-    }
+  async cachedCookAsync(text, options = {}) {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:410]: Unrelated refactoring. Feels like the `async` / `await` reads better/easier than the `Promise`-based version.

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 9]
```diff
@@ -413,92 +407,82 @@ export default Component.extend(TextareaTextManipulation, {
...
+    this.set("preview", cooked);
 
-      this.set("preview", cooked);
+    if (this.siteSettings.enable_diffhtml_preview) {
+      const cookedElement = document.createElement("div");
+      cookedElement.className = this._previewElement.className;
+      cookedElement.innerHTML = cooked;
 
-      let previewPromise = Promise.resolve();
+      // Same order of operation as in the "previewUpdated" method in "composer-editor.js"
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:443]: Kept the same order of operation than in the `previewUpdated` method in the `composer-editor.js` file to be more consistent.

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 10]
```diff
@@ -836,13 +820,13 @@ export default Component.extend(TextareaTextManipulation, {
   },
 
   _disablePreviewTabIndex() {
-    const observer = new MutationObserver(function () {
+    const observer = new MutationObserver(() => {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:823]: Unrelated refactoring.

File: app/assets/javascripts/discourse/app/lib/public-js-versions.js [hunk 11]
```diff
@@ -9,7 +9,6 @@ export const PUBLIC_JS_VERSIONS = {
   "chart.min.js": "chart.js/3.5.1/chart.min.js",
   "chartjs-plugin-datalabels.min.js":
     "chartjs-plugin-datalabels/2.2.0/chartjs-plugin-datalabels.min.js",
-  "diffhtml.min.js": "diffhtml/1.0.0-beta.30/diffhtml.min.js",
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/public-js-versions.js:12]: `diffhtml` is no longer vendored.

File: lib/tasks/javascript.rake [hunk 12]
```diff
@@ -73,7 +73,6 @@ def dependencies
     },
     { source: "chart.js/dist/chart.min.js", public: true },
     { source: "chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.min.js", public: true },
-    { source: "diffhtml/dist/diffhtml.min.js", public: true },
```
Comment (ZogStriP) [inline lib/tasks/javascript.rake:76]: `diffhtml` is no longer vendored.

File: package.json [hunk 13]
```diff
@@ -20,7 +20,6 @@
     "chrome-launcher": "^1.1.0",
     "chrome-remote-interface": "^0.33.0",
     "concurrently": "^8.2.2",
-    "diffhtml": "1.0.0-beta.30",
```
Comment (ZogStriP) [inline package.json:23]: `diffhtml` is no longer vendored.

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 14]
```diff
@@ -223,6 +222,14 @@ export function onToolbarCreate(func) {
   addToolbarCallback(func);
 }
 
+const COMPOSER_EVENTS = [
```
[the remark this answers is not in the document]
Comment (davidtaylorhq) [reply app/assets/javascripts/discourse/app/components/d-editor.js:225]: I realise it's tricky now, but it would be way easier to review (and `git blame` later) if any unrelated refactoring was split into a separate PR. A common technique I use is to stack the PRs: make a refactoring branch, and then base your feature branch on that.

Then we can review/merge the no-op refactor first, rebase the feature PR on `main`, and then review/merge the feature.

File: lib/tasks/javascript.rake (hunk 12 above)
Comment (davidtaylorhq) [reply lib/tasks/javascript.rake:76]: Have we updated all plugins that were using it? (just discourse-ai?)

I wonder if we need to defer 'dropping diffhtml' to a later PR?

File: yarn.lock [hunk 15]
```diff
@@ -5361,7 +5361,7 @@ diff@^5.1.0:
   resolved "https://registry.yarnpkg.com/diff/-/diff-5.1.0.tgz#bc52d298c5ea8df9194800224445ed43ffc87e40"
   integrity sha512-D+mk+qE8VC/PAUrlAU34N+VfXev0ghe5ywmpqrawphmVZc1bEfn56uo9qpyGp1p4xpzOHkSW4ztBd6L7Xx4ACw==
 
-diffhtml@1.0.0-beta.30:
+diffhtml@^1.0.0-beta.30:
```
Comment (davidtaylorhq) [inline yarn.lock:5364]: Huh! How come it stayed in the lockfile? Is it still in one of our `package.json` files? Or does something else depend on it?

Comment (CvX) [reply yarn.lock:5364]: I think it just needs a `yarn` run 😃

Comment (CvX) [reply yarn.lock:5364]: ah, no 😄 it's in `app/assets/javascripts/discourse/package.json` now 😅

File: app/assets/javascripts/discourse/app/components/d-editor.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (ZogStriP) [reply app/assets/javascripts/discourse/app/components/d-editor.js]: Removed all unrelated refactorings.

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 16]
```diff
@@ -413,92 +412,79 @@ export default Component.extend(TextareaTextManipulation, {
     return toolbar;
   },
 
-  cachedCookAsync(text) {
-    if (this._cachedCookFunction) {
-      return Promise.resolve(this._cachedCookFunction(text));
+  async cachedCookAsync(text, options = {}) {
```
[the remark this answers is not in the document]
Comment (ZogStriP) [reply app/assets/javascripts/discourse/app/components/d-editor.js:415]: Right, but not having a default might break other callers expecting only 1 argument. Though, there might not be other callers.

[the remark this answers is not in the document]
Comment (CvX) [reply app/assets/javascripts/discourse/app/components/d-editor.js:415]: it's js! there are no rules! 🤪 (or required arguments 😉)
